orca/src/ds.rs

```diff
@@ -0,0 +1,414 @@
+        let ldap = LdapClient::new(uri, basedn, LdapSchema::Rfc2307bis)?;
+
+        Ok(DirectoryServer { ldap, dm_pw })
+    }
+
+    pub fn build(uri: String, dm_pw: String, basedn: String) -> Result<TargetServer, ()> {
+        Self::construct(uri, dm_pw, basedn).map(TargetServer::DirSrv)
+    }
+
+    #[allow(clippy::new_ret_no_self)]
+    pub fn new(lconfig: &DsConfig) -> Result<TargetServer, ()> {
+        Self::construct(
+            lconfig.uri.clone(),
+            lconfig.dm_pw.clone(),
+            lconfig.base_dn.clone(),
+        )
+        .map(TargetServer::DirSrv)
+    }
+
+    pub fn info(&self) -> String {
+        format!("Directory Server Connection: {}", self.ldap.uri)
+    }
+
+    pub fn builder(&self) -> TargetServerBuilder {
+        TargetServerBuilder::DirSrv(
+            self.ldap.uri.clone(),
+            self.dm_pw.clone(),
+            self.ldap.basedn.clone(),
+        )
+    }
+
+    pub async fn open_admin_connection(&self) -> Result<(), ()> {
+        self.ldap.open_dm_connection(&self.dm_pw).await
+    }
+
+    pub async fn setup_admin_delete_uuids(&self, targets: &[Uuid]) -> Result<(), ()> {
+        // We might hit admin limits depending on the dataset size, so we probably
+        // need to do this iteratively eventually. Or just change the limits ...
+
+        let filter = LdapFilter::Or(
+            targets
+                .iter()
+                .map(|u| LdapFilter::Equality("cn".to_string(), u.to_string()))
+                .collect(),
+        );
+
+        print!("(|");
+        for u in targets.iter() {
+            print!("(cn={})", u);
+        }
+        println!(")");
+
+        let res = self.ldap.search(filter).await?;
+
+        for ent in res.iter() {
+            debug!("Deleting ... {}", ent.dn);
+            self.ldap.delete(ent.dn.clone()).await?;
+        }
+        Ok(())
+    }
+
+    pub async fn setup_admin_precreate_entities(
+        &self,
+        targets: &HashSet<Uuid>,
+        all_entities: &HashMap<Uuid, Entity>,
+    ) -> Result<(), ()> {
+        // Check if ou=people and ou=group exist
+        let res = self
+            .ldap
+            .search(LdapFilter::Equality("ou".to_string(), "people".to_string()))
+            .await?;
+
+        if res.is_empty() {
+            // Doesn't exist
+            info!("Creating ou=people");
+            let ou_people = LdapAddRequest {
+                dn: format!("ou=people,{}", self.ldap.basedn),
+                attributes: vec![
+                    LdapAttribute {
+                        atype: "objectClass".to_string(),
+                        vals: vec!["top".to_string(), "organizationalUnit".to_string()],
+                    },
+                    LdapAttribute {
+                        atype: "ou".to_string(),
+                        vals: vec!["people".to_string()],
+                    },
+                ],
+            };
+            self.ldap.add(ou_people).await?;
+        }
+
+        let res = self
```

Comment:
Why do we need to run this twice?

Reply:
They search different things - one is searching for ou=people, the other, ou=groups.